runtime/lua/vim/provider/python.lua

```diff
@@ -0,0 +1,163 @@
+local M = {}
+local min_version = '3.7'
+local max_version = '3.12'
+local s_err = ''
+local s_host = ''
+
+local python_candidates = {
+  'python3',
+  'python3.12',
+  'python3.11',
+  'python3.10',
+  'python3.9',
+  'python3.8',
+  'python3.7',
+  'python',
+}
+
+--- @param prog string
+--- @param module string
+--- @return integer, string
+local function import_module(prog, module)
+  local program = [[
+import sys, importlib.util;
+sys.path = [p for p in sys.path if p != ""];
+sys.stdout.write(str(sys.version_info[0]) + "." + str(sys.version_info[1]));]]
+
+  program = program
+    .. string.format('sys.exit(2 * int(importlib.util.find_spec("%s") is None))', module)
+
+  local out = vim.system({ prog, '-W', 'ignore', '-c', program }):wait()
+  return out.code, assert(out.stdout)
+end
+
+--- @param prog string
+--- @param module string
+--- @return integer, string
+local function check_for_module(prog, module)
+  local prog_path = vim.fn.exepath(prog)
+  if prog_path == '' then
+    return 0, prog .. ' not found in search path or not executable.'
+  end
+
+  --   Try to load module, and output Python version.
+  --   Exit codes:
+  --     0  module can be loaded.
+  --     2  module cannot be loaded.
+  --     Otherwise something else went wrong (e.g. 1 or 127).
+  local prog_exitcode, prog_version = import_module(prog, module)
+  if prog_exitcode == 2 or prog_exitcode == 0 then
+    -- Check version only for expected return codes.
+    if vim.version.lt(prog_version, min_version) then
+      return 0,
+        string.format(
+          '%s is Python %s and cannot provide Python >= %s.',
+          prog_path,
+          prog_version,
+          min_version
+        )
+    elseif vim.version.gt(prog_version, max_version) then
+      return 0,
+        string.format(
+          '%s is Python %s and cannot provide Python <= %s.',
+          prog_path,
+          prog_version,
+          max_version
+        )
+    end
+  end
+
+  if prog_exitcode == 2 then
+    return 0, string.format('%s does not have the "%s" module.', prog_path, module)
+  elseif prog_exitcode == 127 then
+    -- This can happen with pyenv's shims.
+    return 0, string.format('%s does not exist: %s', prog_path, prog_version)
+  elseif prog_exitcode ~= 0 then
+    return 0,
+      string.format(
+        'Checking %s caused an unknown error. (%s, output: %s) Report this at https://github.com/neovim/neovim',
+        prog_path,
+        prog_exitcode,
+        prog_version
+      )
+  end
+
+  return 1, ''
+end
+
+--- @param module string
+--- @return string[]
```

Comment:
Note: `vim.provider.python.detect_by_module('pynvim')` would return:
```lua
{ "/path/to/valid/.venv/bin/python3", "" }  -- ok
{ "", "ERROR: ..." }  -- error
```

because this function is a direct translation of the vimscript autoload function.

```suggestion
--- @return {[1]: string, [2]: string}
```
where `[1]` is the path to detected python, `[2]` is an error.

Or a data structure will be better (the vimscript bridge also needs to be adjusted accordingly):

```suggestion
--- @return { ['executable']: string, ['error']: string }
```
(or equivalent `@class` `@field` etc.)

But I think a more Lua-idomatic way would be much better, since this is a new API:
```suggestion
--- @return executable string?  path to detected python, if any; nil or not found
--- @return error string? error message if python can't be detected by {module}; nil if success
```